1_nwisdata/src/nwis_pull.R

```diff
@@ -0,0 +1,308 @@
+
+  nwis_state_codes <- dataRetrieval::stateCd
+
+  # read in the inventory, which includes all dates and is therefore a superset
+  # of what we'll be pulling for a specific date range
+  scipiper::sc_retrieve(inventory_ind)
+  inventory <- feather::read_feather(scipiper::as_data_file(inventory_ind)) %>%
+    select(Constituent, Site=MonitoringLocationIdentifier, StateCode, resultCount)
+
+  # split up the work into manageably sized partitions (~max size specified in nwis_pull)
+  partitions <- bind_rows(lapply(unique(inventory$Constituent), function(constituent) {
+    bind_rows(lapply(unique(inventory$StateCode), function(statecode) {
+      # an atomic group is a combination of parameters that can't be reasonably
+      # split into multiple nwis pulls - in this case we're defining atomic
+      # groups as distinct combinations of constituent (a group of
+      # characteristicNames) and site ID. We could potentially split
+      # constituents further, or pull multiple constituents at once, or pull an
+      # entire state at once, but we're doing it this way
+      atomic_groups <- inventory %>%
+        filter(StateCode==statecode, Constituent == constituent) %>%
+        group_by(Site) %>%
+        summarize(NumObs=sum(resultCount)) %>%
+        arrange(desc(NumObs))
+
+      # split the full pull (combine atomic groups) into right-sized partitions
+      # by an old but fairly effective paritioning heuristic: pick the number of
+      # partitions desired, sort the atomic groups by descending size, and then
+      # go down the list, each time adding the next atomic group to the
+      # partition that's currently smallest
+      num_partitions <- ceiling(sum(atomic_groups$NumObs) / nwis_pull$target_pull_size)
+      partition_sizes <- rep(0, num_partitions)
+      assignments <- rep(0, nrow(atomic_groups))
+      for(i in seq_len(nrow(atomic_groups))) {
+        size_i <- atomic_groups[[i,"NumObs"]]
+        smallest_partition <- which.min(partition_sizes)
+        assignments[i] <- smallest_partition
+        partition_sizes[smallest_partition] <- partition_sizes[smallest_partition] + size_i
+      }
+
+      # create and return data.frame rows for binding in the two lapply loops
+      state <- nwis_state_codes %>%
+        filter(value==sprintf('US:%s', statecode)) %>%
+        pull(name)
+      atomic_groups %>%
+        mutate(
+          State=state,
+          StateCode=statecode,
+          Constituent=constituent,
+          NumPulls=num_partitions,
+          PullTask=sprintf('%s_%s_%03d', gsub(' ', '_', state), constituent, assignments))
+    }))
+  }))
+
+  # prepare nested collections of other possible arguments to include in the readNWISdata calls
+  parameter_codes <- bind_rows(lapply(unique(partitions$Constituent), function(constituent) {
+    data_frame(Constituent=constituent, CharacteristicNames=nwis_codes$characteristicName[[constituent]]) %>%
+      tidyr::nest(-Constituent, .key='Params')
```

Comment:
I've never seen or used `nest` before, and just explored it a bit - so cool! I immediately see an lapply+bind_rows combo that I recently did being fixed by `unnest()`